include/swift/Sema/ConstraintSystem.h

```diff
@@ -2157,6 +2157,11 @@ class ConstraintSystem {
   /// Note: this is only used to support ObjCSelectorExpr at the moment.
   llvm::SmallPtrSet<Expr *, 2> UnevaluatedRootExprs;
 
+  /// Expressions that are in the left-hand side of an assignment
+  /// expression. This is used to help distinguish assignments from
+  /// any other potential use that might need lvalues.
+  llvm::SmallPtrSet<Expr *, 2> TargetOfAssignExprs;
```

Comment:
I think this needs to be controlled by the solver scope because result builder generate assignments as part of the transform which means that if not controlled this can grow pretty large.

Reply:
Since the constraint system records parent chains maybe this could be replaced with a walk up the parent chain until it reaches assignment expression?

Reply:
I missed the fact that we record parent chains; I'd much prefer that to having to keep solver state!

Reply:
I did this and it's so much better, thank you!